cni/pkg/iptables/iptables.go

```diff
@@ -0,0 +1,458 @@
+	if err := cfg.executeCommands(builder); err != nil {
+		log.Errorf("failed to add host netnamespace iptables rules: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (cfg *IptablesConfigurator) DeleteHostRules() {
+	log.Debug("Attempting to delete hostside iptables rules (if they exist)")
+
+	cfg.executeHostDeleteCommands()
+}
+
+func (cfg *IptablesConfigurator) executeHostDeleteCommands() {
+	optionalDeleteCmds := [][]string{
+		// delete our main jump in the host ruleset. If it's not there, NBD.
+		{"-t", iptablesconstants.NAT, "-D", iptablesconstants.POSTROUTING, "-j", ChainHostPostrouting},
+		// flush-then-delete our created chains
+		// these sometimes fail due to "Device or resource busy" - again NBD.
+		{"-t", iptablesconstants.NAT, "-F", ChainHostPostrouting},
+		{"-t", iptablesconstants.NAT, "-X", ChainHostPostrouting},
+	}
+
+	// iptablei seems like a reasonable pluralization of iptables
+	iptablei := []string{iptablesconstants.IPTABLES}
+	if cfg.cfg.EnableInboundIPv6 {
+		iptablei = append(iptablei, iptablesconstants.IP6TABLES)
+	}
+	for _, iptables := range iptablei {
+		for _, cmd := range optionalDeleteCmds {
+			err := cfg.ext.Run(iptables, nil, cmd...)
+			if err != nil {
+				log.Debugf("ignoring error deleting optional iptables rule: %v", err)
+			}
+		}
+	}
+}
+
+func (cfg *IptablesConfigurator) appendHostRules(hostSNATIP *netip.Addr) *builder.IptablesBuilder {
+	log.Info("configuring host-level iptables rules (healthchecks, etc)")
+
+	iptablesBuilder := builder.NewIptablesBuilder(ipbuildConfig(cfg.cfg))
+
+	// For easier cleanup, insert a jump into an owned chain
+	// -A POSTROUTING -p tcp -j ISTIO_POSTRT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.POSTROUTING, iptablesconstants.NAT,
+		"-j", ChainHostPostrouting,
+	)
+
+	// TODO BML I don't think we need UDP? TCP healthcheck redir should catch everything.
+
+	// This is effectively an analog for Istio's old-style podSpec-based health check rewrites.
+	// Before Istio would update the pod manifest to rewrite healthchecks to go to sidecar Envoy port 15021,
+	// so that it could distinguish things that can be unauthenticated (healthchecks) from other kinds of node traffic
+	// (e.g. LoadBalanced Service packets, etc) that need to be authenticated/captured/proxied.
+	//
+	// We want to do the same thing in ambient but can't rely on podSpec injection. So, do effectively the same thing,
+	// but with iptables rules - use `--socket-exists` as a proxy for "is this a forwarded packet" vs "is this originating from
+	// a local node socket". If the latter, outside the pod in the host netns, redirect that traffic to a hardcoded/custom proxy
+	// healthcheck port, just like we used to. Otherwise, we can't assume it's local-node privileged traffic, and will capture and process it normally.
+	//
+	// All this is necessary because quite often apps use the same port for healthchecks as they do for reg. traffic, and
```

Comment:
How does this not make the entire check unreliable? We are treating ports specially for reasons broader than they are intended by the user

Reply:
It's the same check we do in ambient CNI today (slightly stricter in fact).

- In current ambient CNI we let all traffic thru if it comes from the node (as determined by `--socket-exists`), regardless of port (i.e. healthcheck ports and everything else, inclusive).

- In this ambient CNI we let all traffic thru if it comes from the node (as determined by `--socket-exists`), only if it is destined for something explicitly declared as a healthcheck port by the pod manifest.


So, it's strictly better than the current ambient impl, and the best we can do short of rewriting people's pod manifests for them (which we'd rather not do, since that requires a webhook, and is generally not great).

The other option is to proxy node healthchecks thru ztunnel, which could be done but isn't really necessary/desirable.

Reply:
We are SNATing it yes, that's not likely to cause any issues. DNATing would, which is why we aren't doing that.

Outside of the scenario where an application pod expects a a specific fixed source IP for nodes that is not the one we're SNATing to, which we can work around pretty simply, there aren't any other valid scenarios we could come up with where this would problematically break. 

It is also strictly better than current code in terms of ingress control.

If you have specific articulated scenarios were SNATing node IP for kubelet healthchecks breaks down that aren't already covered, we'd like to hear them - they may exist, we just couldn't think of or find any.

> I think special casing health check ports is going to lead to weird behavior.

I agree. Istio has always special cased them, however, and it has always led to weird behavior of one kind or another, so for the short term we are likely stuck with it. Long-term, without establishing kubelet source identity or pod manifest rewrites (itself a form of weird behavior), this is not fixable in a more reliable, less intrusive way.

Reply:
If SNAT is safe and not going to break anything, IMO we should do it unconditionally for this host traffic, not just for certain ports. If its _not safe_, then its probably not appropriate for a health check scenario?

Reply:
Sure, we could SNAT everything originating from the host node whether it was destined to health-check ports or not, on the assumption that it doesn't matter because if the host node is compromised and running hostile processes, they could be "smart" enough to just target health check ports anyway.

 Why though? We literally don't _need_ to do it for anything but healthcheck ports.

Reply:
Consistency is important. Is it weird that we SNAT host traffic? Yes... but at least we can be consistently weird. Special casing a port because it happens to be in the health check is bizarre behavior that will be inscrutable to users and lead to hard-to-understand behavior.

Reply:
> Special casing a port because it happens to be in the health check is bizarre behavior that will be inscrutable to users and lead to hard-to-understand behavior.

Yes, the part I'm hung up on is why it becomes somehow less inscrutable and bizarre if we do it for everything, even the ports where it's not strictly required and indicated with user intent in a declarative manifest.

Do you have some concrete examples, maybe?

Reply:
Sure. Consider I have some workload that is getting host traffic on port 9090. It calls my :9090/some/api. For whatever reason, I am looking at the source IP -- maybe its just logs for debugging, network tracing tools, whatever.

Now I add a liveness probe on :9090/healthz for operational reasons.

no user would expect these two things, which are entirely unrelated ,to impact each other. However, because we make an arbitrary leap between networking path on the entire port, all of port 9090 is impacted.

Later, Ichange my liveness probe to port 8080. Again, things change for reasons that are inscrutable to users.

Principal of least surprise

Reply:
Sure, that is surprising.

It's not *more* surprising than ignoring user-declared healthcheck ports in the manifest and allowing unauthenticated node traffic to hit *all* ports of the pod, when that's explicitly not what the user asked for.

Today, if I do not want to allow _any_ unauthenticated host node traffic to hit my pod, I can simply not specify health check ports, and it will be blocked, which is what I would expect, since I defined none.

If we ignore healthcheck ports and do that for all node traffic, it would not be blocked, which would be both surprising and concerning, and expressly not respecting the declarative pod manifest.

Reply:
I am very much in favor of consistent behavior, especially around traffic allow vs deny behavior.

Setting a readiness probe **does not** mean to allow all traffic on that port.

Its 100% fine to say "Istio ambient doesn't block/handle host traffic at all". its not great to say "Istio ambient doesn't block host traffic in some weird cases". It would be terrible to say "Istio ambient doesn't block host traffic in random cases" (just as an example -- obviously we are not doing that).

This reminded me I actually have a draft blog post I have been meaning to dust off that basically says this. Basically in security policies, its more important to be consistent than to be comprehensive; if there are known areas they do not cover its easy to workaround, but if they coverage is inconsistent or hard to understand its hard to operate. A good example is Istio not having egress policies - very easy to workaround as a user. A bad example would be for Istio to not enforce policy during node startup - very hard to workaround. This is somewhere in between IMO.

Reply:
> Its 100% fine to say "Istio ambient doesn't block/handle host traffic at all". its not great to say "Istio ambient doesn't block host traffic in some weird cases". It would be terrible to say "Istio ambient doesn't block host traffic in random cases" (just as an example -- obviously we are not doing that).

The problem is that ignoring all "host traffic" also leads to inconsistency, because there are two kinds of "host traffic" (as mentioned in https://github.com/istio/istio/pull/46936)


- traffic originating from sockets on the host (kubelet) - that is, traffic from processes on the node -> this will enter the pod with HostIP as src.
- traffic originating from outside the cluster (kube-proxy, load balancer-backed services, etc) - that is, traffic from outside the node -> this will also enter the pod with HostIP as src.

If we simply skip all traffic from hostIP, we let traffic from the latter bypass all ambient authz rules, which seems to be a pretty big hole - I think that would be pretty surprising to most people.

There's also the fact that our previous attempts to bypass host IP traffic have been pretty hacky with lots of edge cases - actually figuring out what "the host IP" will be from the perspective of a given pod is _not_ trivial, or even possible to do reliably in all cases, and we have had bugs around this for years - that's why this doesn't bother to try to figure out what the host IP is at all, which is much simpler/more reliable.

Reply:
> The problem is that ignoring all "host traffic" also leads to inconsistency,

I think this is just a terminology issue. Feel free to replace anywhere i used "host traffic" with "traffic originating from sockets on the host"

Reply:
Briefly discussed in ambient meeting: https://docs.google.com/document/d/1SMlwliEnthgq7r2PjpLl1kCq3t8rAMbgu6r_lDAXJ0w/edit#heading=h.dmh9c2nz99ep

I'll keep the SNAT but remove the healthcheck port logic bits.

Reply:
So pod with hostnetwork can access the server without any auth check.

Reply:
yes, all processes in the host netns will have direct access to *local* pods. Otherwise we break k8s health checks.

Reply:
One thought to iterate here by allowing the user to specify kubelet's UID, and write the iptables rules only allowing that UID. i don't think we can know this ahead of time, so this would have to be user input at install time... So current behaviour will still be the default.

Reply:
There's no way around the fact that someone putting healthCheck ports in their pod manifest is express intent that they want to allow unauthenticated node-process traffic to hit their app.

That's just what healthChecks are in Kubernetes today. If you allow them at all, you have this problem.

Reply:
We could simply live with that and only add the general host traffic exclusion rule if at least one health check port is defined, so that if you explicitly didn't want host traffic you have an out - just don't add healthchecks.

But to John's original point, that would be an Istio opinion, not a Kube one.